theano/gpuarray/subtensor.py

```diff
@@ -800,7 +800,7 @@ def make_node(self, x, y, ilist):
         """
         ctx_name = infer_context_name(x, y, ilist)
         x_ = as_gpuarray_variable(x, ctx_name)
-        y_ = as_gpuarray_variable(y, ctx_name)
+        y_ = as_gpuarray_variable(y.astype(x.dtype), ctx_name)
```

Comment:
Why not cast it inside the c code before calling atomicAdd or atomicExch around line 1000 in this file? It would  prevent a copy in memory and seem easy to do.